src/wast-parser.cc

```diff
@@ -1324,8 +1293,53 @@ bool WastParser::PeekIsCustom() {
          tokens_.front().text() == "custom";
 }
 
+Result WastParser::ResolveRefTypes(const Module* module,
+                                   TypeVector& types,
+                                   ReferenceVars& ref_vars,
+                                   Errors* errors) {
+  Result result = Result::Ok;
+
+  for (auto& ref_var : ref_vars) {
+    uint32_t index = ref_var.index;
+
+    // The index of resolved variables is converted to kInvalidIndex.
+    if (index == kInvalidIndex) {
+      continue;
+    }
+
+    assert(index < types.size());
+    Type type = types[index];
+    ref_var.index = kInvalidIndex;
+
+    assert(type.IsReferenceWithIndex());
+
+    if (type.GetReferenceIndex() != kInvalidIndex) {
+      continue;
+    }
+
+    const auto type_index = module->type_bindings.FindIndex(ref_var.var.name());
+
+    if (type_index == kInvalidIndex) {
+      errors->emplace_back(ErrorLevel::Error, ref_var.var.loc,
+                           StringPrintf("undefined reference type name %s",
+                                        ref_var.var.name().c_str()));
+      result = Result::Error;
+      continue;
+    }
+
+    types[index] = Type(static_cast<Type::Enum>(type), type_index);
+  }
+
+  return Result::Ok;
+}
+
 Result WastParser::ParseModuleFieldList(Module* module) {
   WABT_TRACE(ParseModuleFieldList);
+
+  // Parser may parse multiple modules. Reset internal status.
+  resolve_types_.clear();
+  resolve_funcs_.clear();
```

Comment:
There seems to be whole bunch of other state stored in `WastParser` (e.g. errors_, last_module_index_, etc)  .. are you sure re-use is supported?  Or is the other state reset somewhere else?

Reply:
I don't know. A spec test may contain several modules,

This is an example call stack:
```
#0  wabt::WastParser::ParseModuleFieldList
#1  wabt::WastParser::ParseScriptModule
#2  wabt::WastParser::ParseModuleCommand
#3  wabt::WastParser::ParseCommand
#4  wabt::WastParser::ParseCommandList
#5  wabt::WastParser::ParseScript
#6  wabt::ParseWastScript
```

The `ParseCommandList` has a loop. But nothing looks like a reset. I suspect the new module construction is the "reset". The point of this patch is moving out parser related stuff from the module, so I suspect this is the reason of the missing reset. Any suggestion?

Reply:
I see, so the rest of the state is not per-module state but per-wast-file state.

In that case I agree this is probably needed.   Maybe  update the commend to say something like "Reset module-specific state"?

Reply:
Patch updated. Thank you!